### PR TITLE
Student/Staff pages #107/#108

### PR DESCRIPTION
**Pull request by jonahchoi**
#107 #108 : Added table component to render students and staff.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -17,8 +17,13 @@
     "lint:fix": "eslint . --fix"
   },
   "dependencies": {
-    "@aws-sdk/client-s3": "^3.360.0",
-    "@aws-sdk/s3-request-presigner": "^3.360.0",
+    "@aws-sdk/client-s3": "^3.341.0",
```
Comment (amantri) [inline package.json:20]:
```suggestion
    "@aws-sdk/client-s3": "^3.360.0",
```

File: package.json [hunk 2]
```diff
@@ -17,8 +17,13 @@
...
+    "@aws-sdk/s3-request-presigner": "^3.341.0",
```
Comment (amantri) [inline package.json:21]:
```suggestion
    "@aws-sdk/s3-request-presigner": "^3.360.0",
```

File: package.json (hunk 1 above)
Comment (amantri) [reply package.json:20]: Note: you will have to update package-lock.json in this PR after this change.

File: src/components/case_manager/MyParas.tsx [hunk 3]
```diff
@@ -41,24 +40,68 @@ const MyParas = () => {
...
+      label: "Last Name",
+      hasInput: true,
+    },
+    {
+      id: "email",
+      label: "Email",
+      hasInput: true,
+    },
+    {
+      id: "phone",
```
Comment (amantri) [inline src/components/case_manager/MyParas.tsx:60]: Let's only keep the first name, last name and email columns for now. We can add more columns to this table over time as we start collecting more data.

File: src/components/case_manager/MyParas.tsx [hunk 4]
```diff
@@ -41,24 +40,68 @@ const MyParas = () => {
...
+  return (
+    <>
+      <PersonTable
+        people={paras as Para[]}
+        onSubmit={handleSubmit}
+        headCells={headCells}
+        type="Staff"
+      />
+    </>
+    // <div>
```
Comment (amantri) [inline src/components/case_manager/MyParas.tsx:90]: Please delete these lines

File: src/components/case_manager/MyStudents.tsx [hunk 5]
```diff
@@ -28,25 +27,40 @@ const MyStudentsPage = () => {
...
+      label: "Last Name",
+      hasInput: true,
+    },
+    {
+      id: "email",
+      label: "Email",
+      hasInput: true,
+    },
+    {
+      id: "dateAdded",
```
Comment (amantri) [inline src/components/case_manager/MyStudents.tsx:47]: Ditto - remove this column for now.

File: src/components/case_manager/types/table.ts [hunk 6]
```diff
@@ -0,0 +1,45 @@
+export interface Person {
+  first_name: string;
+  last_name: string;
+  email: string;
+  phone?: string;
```
Comment (amantri) [inline src/components/case_manager/types/table.ts:5]: Remove phone and dateAdded for now. In general, it is easier to add fields/data to interfaces later than to remove, so err on the side of minimalism.

Also remove other non-essential fields in the other interfaces

File: src/components/case_manager/types/table.ts [hunk 7]
```diff
@@ -0,0 +1,45 @@
+export interface Person {
```
Comment (amantri) [inline src/components/case_manager/types/table.ts:1]: It would be useful to have these interfaces in a more central location. 
@codetheweb: any ideas on where to put them?

File: src/components/para_trials/timer.tsx
Comment (amantri) [inline src/components/para_trials/timer.tsx:1]: Move this and related files to a different PR?

File: src/components/case_manager/types/table.ts (hunk 7 above)
Comment (jonahchoi) [reply src/components/case_manager/types/table.ts:1]: Fixed everything except this one. We can either wait for a response from Max or change it later. What do you think?

**Review by amantri — APPROVED** [review]

File: src/components/case_manager/MyParas.tsx [hunk 8]
```diff
@@ -41,24 +40,33 @@ const MyParas = () => {
     return <div>Loading...</div>;
   }
 
-  return (
-    <div>
-      <PersonCreationForm title={"Add a Para"} onSubmit={handleSubmit} />
+  const headCells: HeadCell[] = [
```
[the remark this answers is not in the document]
Comment (jonahchoi) [reply src/components/case_manager/MyParas.tsx:43]: No, its just what MUI had defined in their example, which I copied over from the docs. Since the entries are called TableCell, they used HeadCell for the header. I'm fine with renaming it if you prefer.

File: src/components/case_manager/PersonTable.tsx [hunk 9]
```diff
@@ -0,0 +1,445 @@
...
+  order: Order,
+  orderBy: UserKeys
+): <T extends Student | Para>(a: T, b: T) => number {
+  return order === "desc"
+    ? (a, b) => descendingComparator(a, b, orderBy)
+    : (a, b) => -descendingComparator(a, b, orderBy);
+}
+
+// Since 2020 all major browsers ensure sort stability with Array.prototype.sort().
+// stableSort() brings sort stability to non-modern browsers (notably IE11). If you
```
[the remark this answers is not in the document]
Comment (jonahchoi) [reply src/components/case_manager/PersonTable.tsx:48]: Sounds good! I'll remove stableSort and replace it with Array.sort.

[the remark this answers is not in the document]
Comment (codetheweb) [reply src/components/case_manager/PersonTable.tsx:48]: The bundling built into Next.js already adds some polyfills, and we could support IE if we really wanted by changing a config parameter: https://nextjs.org/docs/architecture/supported-browsers

In general we can use the most recent ES features and not worry about compatibility.

File: src/components/case_manager/types/table.ts (hunk 7 above)
Comment (codetheweb) [reply src/components/case_manager/types/table.ts:1]: all the types in this file are already present in `src/backend/db/zapatos/schema.d.ts`

we can remap them slightly if necessary or re-export them to make things more ergonomic, but we need to stick with a single source of truth for database types

for example:
```ts
import { SelectableForTable } from "zapatos/schema"
type User = SelectableForTable<"user">;
```

**Review by codetheweb — CHANGES_REQUESTED** [review]

File: src/components/case_manager/types/table.ts (hunk 7 above)
Comment (jonahchoi) [reply src/components/case_manager/types/table.ts:1]: Thanks for pointing that out. Didn't realize those existed already. I'll likely import them into this file, so that the HeadCell interface can still access them, and re-export for the Table.

File: src/components/case_manager/types/table.ts [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+export interface StudentHeadCell extends HeadCell {
+  id: keyof Student;
+}
+
+export interface ParaHeadCell extends HeadCell {
+  id: keyof Para;
+}
+
+export function isStudent(person: Student | Para): person is Student {
+  return (person as Student).student_id !== undefined;
```
Comment (codetheweb) [inline src/components/case_manager/types/table.ts:21]: nice

File: src/components/case_manager/PersonTable.tsx [hunk 11]
```diff
@@ -0,0 +1,398 @@
...
+            {showInput && (
+              <EnhancedTableInput
+                inputCells={headCells}
+                type={type}
+                onCloseInput={handleCloseInput}
+              />
+            )}
+            {visibleRows.map((row, index) => {
+              const isItemSelected = isSelected(row.email);
+              const labelId = `enhanced-table-checkbox-${index}`;
```
Comment (codetheweb) [inline src/components/case_manager/PersonTable.tsx:355]: 🐧 is the prefix here used for anything?
```suggestion
              const labelId = row.email;
```

File: src/components/case_manager/PersonTable.tsx [hunk 12]
```diff
@@ -0,0 +1,398 @@
...
+      <EnhancedTableToolbar
+        numSelected={selected.length}
+        totalRows={people.length}
+        type={type}
+        onOpenInput={handleOpenInput}
+      />
+      <TableContainer>
+        <Table
+          sx={{ minWidth: 750 }}
+          aria-labelledby={`Table of ${"students/paras"}`}
```
Comment (codetheweb) [inline src/components/case_manager/PersonTable.tsx:334]:
```suggestion
          aria-labelledby={`Table of students/paras`}
```

File: src/components/case_manager/PersonTable.tsx [hunk 13]
```diff
@@ -0,0 +1,398 @@
...
+  "&.Mui-selected": {
+    backgroundColor: "lightblue",
+  },
+}));
+
+interface EnhancedTableProps<Person, Column> {
+  people: Person[];
+  onSubmit: (event: React.FormEvent<HTMLFormElement>) => void;
+  headCells: Column[];
+  type: "Student" | "Staff";
```
Comment (codetheweb) [inline src/components/case_manager/PersonTable.tsx:250]: can this be inferred based on `people`?

**Review by codetheweb — APPROVED** [review]

File: src/components/case_manager/PersonTable.tsx (hunk 13 above)
Comment (jonahchoi) [reply src/components/case_manager/PersonTable.tsx:250]: We could use the `isStudent` typeguard function on the first element of the `people` array to define `type`. But the issue is if there are no elements in `people` yet, how would we know to render the students or staff version?
